core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala

```diff
@@ -538,43 +604,56 @@ private[spark] object IndylambdaScalaClosures extends Logging {
 
         override def visitMethodInsn(
             op: Int, owner: String, name: String, desc: String, itf: Boolean): Unit = {
-          if (owner == implClassInternalName) {
-            val ownerExternalName = owner.replace('/', '.')
+          val ownerExternalName = owner.replace('/', '.')
+          if (owner == currentClassInternalName) {
             logTrace(s"    found intra class call to $ownerExternalName.$name$desc")
-            val calleeMethodId = MethodIdentifier(implClass, name, desc)
-            if (!visited.contains(calleeMethodId)) {
-              stack.push(calleeMethodId)
+            // could be invoking a helper method or a field accessor method, just follow it.
+            pushIfNotVisited(MethodIdentifier(currentClass, name, desc))
+          } else if (isInnerClassCtorCapturingOuter(
+              op, owner, name, desc, currentClassInternalName)) {
+            // Discover inner classes.
+            // This this the InnerClassFinder equivalent for inner classes, which still use the
+            // `$outer` chain. So this is NOT controlled by the `findTransitively` flag.
+            logTrace(s"    found inner class $ownerExternalName")
+            val innerClassInfo = getOrUpdateClassInfo(owner)
+            val innerClass = innerClassInfo._1
+            val innerClassNode = innerClassInfo._2
+            trackedClasses += innerClass
+            // We need to visit all methods on the inner class so that we don't missing anything.
+            for (m <- innerClassNode.methods.asScala) {
+              pushIfNotVisited(MethodIdentifier(innerClass, m.name, m.desc))
             }
+          } else if (findTransitively &&
+              trackedClasses.find(_.getName == ownerExternalName).isDefined) {
```

Comment:
This would be cleaner if `trackedClasses` were a `Map` keyed on the class name.

Reply:
Thanks! It would indeed. I had other uses for the trackedClasses but they were gone later and I had forgotten to put this set back...